engine/lib/mySQLRepository.js

```diff
@@ -0,0 +1,108 @@
+'use strict';
+
+const knex = require('knex')
+const logger = require('./logger')
+const fileUtil = require('./fileUtil')
+
+class DatabaseRepository {
+  constructor(databaseConfiguration, invLoader = () => fileUtil.readJsonFile('data/inventory.json')) {
+    this._connection = undefined
+    this._invLoader = invLoader
+
+    const { host, port, user, password, name } = databaseConfiguration
+    this._configuration = {
+      host,
+      port,
+      user,
+      password,
+      database: name,
+      charset: 'utf8'
+    }
+  }
+
+  async isConnected() {
+    if (!this._connection) {
+      await this.initialize()
+    }
+
+    try {
+      await this._connection.select('1')
+      return true
+    }
+    catch (err) {
+      return false
+    }
+  }
+
+  async initialize() {
+    logger.info('Initializing MySQL database')
+    const tableName = 'inventory'
+    await this.createDatabase()
+
+    const connection = knex({
+      client: 'mysql',
+      connection: this._configuration
+    })
+
+    await connection.schema.dropTableIfExists(tableName)
+
+    logger.info('Creating table with name: inventory')
+    await connection.schema.createTable('inventory', table => {
+      table.string('id')
+      table.string('item')
+      table.string('price')
+      table.string('sku')
+      table.string('description')
+    })
+
+    const inventoryData = this._invLoader()
+    for (const item of inventoryData) {
+      logger.info(`Inserting inventory with id: ${item.id}`)
+      await connection(tableName).insert(item)
+    }
```

Comment:
Totally optional, but you could use a `forEach` and some point-free coding here:

```js
const conn = connection(TableName);
const inventoryData = this._invLoader();

inventoryData.forEach(conn.insert);
```

Reply:
Correct me if I'm wrong, but using `forEach` removes the ability to wait for the promises to resolve. Since the database is initialized on the first call to the service, I want to make sure everything is in the table before the first query runs.

Reply:
I found a nice middle ground, unfortunately it doesn't use point-free coding (as I understand it). However, now it will await all the inserts at once instead of blocking on each insert.